numba/typing/templates.py

```diff
@@ -206,6 +206,10 @@ class AbstractTemplate(FunctionTemplate):
     def apply(self, args, kws):
         generic = getattr(self, "generic")
         sig = generic(args, kws)
+        # Enforce that *generic()* must return None or Signature
+        if sig is not None:
+            if not isinstance(sig, Signature):
+                raise AssertionError("{} returned {}".format(self, type(sig)))
```

Comment:
Could this perhaps explain that generic must return None or sig along with the debug info?

Reply:
done